recipe/2024-04-terraform-provider-secrets.md

```diff
@@ -0,0 +1,263 @@
+}
+```
+
+Engine Updates
+
+```go
+//*** CURRENT ***//
+
+type SecretsLoader interface {
+	LoadSecrets(ctx context.Context, secretStore string) (v20231001preview.SecretStoresClientListSecretsResponse, error)
+}
+
+// BaseOptions is the base options for the driver operations.
+type BaseOptions struct {
+  ...
+
+	// Secrets specifies the module authentication information stored in the secret store.
+	Secrets v20231001preview.SecretStoresClientListSecretsResponse
+}
+
+
+//*** NEW ***//
+
+// LoadSecrets function in SecretsLoader interface will populate resolved secrets based on the Ids returned from FindSecretIDs(). The result is passed as BaseOptions.Secrets to the Driver in Execute() call.
+type SecretsLoader interface {
+	LoadSecrets(secretIds map[string][]string, ...)(secretData map[string]map[string]string, error) // secretData is a map of secretStoreID to map of [secretKey]value
+}
+
+// When calling driver.Execute we pass in BaseOptions which is structured as:
+type BaseOptions struct {
+   ...
+
+	// Secrets specifies the module authentication information stored in the secret store.
+	Secrets map[string]map[string]string //Secrets is a map of secretStoreID to map of [secretKey]value
+}
+```
+
+### Error Handling
+
+If there is an error when iterating through the recipe configuration or calling ListSecrets API, error will be returned. These errors will not include sensitive data on secret keys or values.
+
+## Test plan
+
+Testing will include unit tests on all new functions created and a
+functional Test will be created to test this functionality e2e.\
+Unit Tests should include test for 'secret not found'
+and verify that upon returning an error, we do not return sensitive information in the error message.
+
+## Security
+
+Secret data will be held in memory and be passed to the Driver which will inject these values into the Terraform configuration in working directory for Terraform. This directory is created when a Terraform recipe is set to be deployed and deleted once deployment of the Terraform recipe is completed.
+
+Secret data will not be persisted and secret data will not be logged by Radius.
+Terraform recipe authors should, in general, set the sensitive flag for any input and output variables that contain sensitive data.
+
+## Compatibility (optional)
+N/A
+
+## Monitoring and Logging
+
+No updates to existing metrics and logs. 
+We need to make sure we do not log sensitive information within calls.
+
+## Development plan
+Updates to Engine, Driver function call - (under 0.5 Sprint)
+Unit Tests + Functional Test - (under 0.5 Sprint)
+
+Total: 1 Sprint
+
+## Open Questions
+
+We are expecting information on detailed scenarios from PMs on client use cases for Recipe Specific Configuration versus Env Level Provider Config which may get us to revisit design around passing secret data to the driver.
```

Comment:
When is this expected? Is this being tracked somewhere?

Reply:
is there a reason why we need recipe specific configs? i.e. i haven't heard of users asking for this, has anyone else heard of demand for this?

Reply:
With the current design we configure TF providers at the env level. During a recipe execution, data will be processed for all these provider configurations including secrets and passed to the Driver.
During design discussions, it was brought up that it's possible for a specific recipe execution to not need all providers configured and therefore all data sent across and it led to possibility of a more granular approach - recipe specific provider configs.
We were looking to get some more input from PMs after discussions with clients on their user scenarios. As you mention, this is currently not being requested, I can move this out of Open Questions to Design Review Notes and we can revisit this in the future based on feedback.